src/staking/manager.ts

```diff
@@ -0,0 +1,660 @@
+import { networks, Psbt, Transaction } from "bitcoinjs-lib";
+import { StakingParams, VersionedStakingParams } from "../types/params";
+import { UTXO } from "../types";
+import { StakerInfo, Staking } from ".";
+import { fromBech32 } from "@cosmjs/encoding";
+import {
+  btccheckpoint,
+  btcstaking,
+  btcstakingtx,
+} from "@babylonlabs-io/babylon-proto-ts";
+import {
+  BTCSigType,
+  ProofOfPossessionBTC,
+} from "@babylonlabs-io/babylon-proto-ts/dist/generated/babylon/btcstaking/v1/pop";
+import { BBN_REGISTRY_TYPE_URLS } from "../constants/bbnRegistry";
+import { createCovenantWitness } from "./transactions";
+import { getBbnParamByBtcHeight, getBbnParamByVersion } from "../utils/staking/param";
+import { reverseBuffer, uint8ArrayToHex } from "../utils";
+
+export interface BtcProvider {
+  // Sign a PSBT
+  signPsbt(psbtHex: string): Promise<string>;
+  // Sign a message using the ECDSA type
+  signMessage: (message: string, type: "ecdsa") => Promise<string>;
+  // Get the staker info
+  getStakerInfo(): Promise<StakerInfo>;
+  // Get the available UTXOs for staking
+  getUTXOs(): Promise<UTXO[]>;
+}
+
+export interface BbnProvider {
+  // Get the Babylon BTC tip height. Note the there might be a delay between
+  // the BTC tip height and the Babylon BTC tip height. You should not use the
+  // BTC tip height to determine the BBN param.
+  getBabylonBtcTipHeight(): Promise<number>;
+  // Get the Babylon Chain address
+  getBbnAddress(): Promise<string>;
+  // Sign a babylon transaction.
+  signTransaction: <T extends object>(msg: {
+    typeUrl: string;
+    value: T;
+  }) => Promise<Uint8Array>
+}
+
+interface StakingInputs {
+  finalityProviderPkNoCoordHex: string;
+  stakingAmountSat: number;
+  stakingTimelock: number;
+}
+
+interface InclusionProof {
+  pos: number;
+  merkle: string[];
+  blockHash: string;
+}
+
+export class BabylonBtcStakingManager {
+  private stakingParams: VersionedStakingParams[];
+  private btcProvider: BtcProvider;
+  private network: networks.Network;
+  private bbnProvider: BbnProvider;
+
+  constructor(
+    network: networks.Network,
+    stakingParams: VersionedStakingParams[],
+    btcProvider: BtcProvider,
+    bbnProvider: BbnProvider,
+  ) {
+    this.network = network;
+    this.btcProvider = btcProvider;
+    this.bbnProvider = bbnProvider;
+
+    if (stakingParams.length === 0) {
+      throw new Error("No staking parameters provided");
+    }
+    this.stakingParams = stakingParams;
+  }
+
+  /**
+   * Creates an signed EOI transaction that is ready to be sent to the BBN chain.
```

Comment:
```suggestion
   * Creates an signed pre-stake registration transaction that is ready to be sent to the BABY chain.
```

I think we should call pre-stake registration instead of EOI, in the sense that you are submitting first to Babylon before staking. It allows to specify for "post-stake registration" as the legacy way to submit txs.

Good for us to agree on the naming conventions as these libraries will be used by externals.